### PR TITLE
Placeorder

### PR DESCRIPTION
**Pull request by Gupta-Shivanee**
Trello Ticket:- https://trello.com/c/4W2DirvJ/19-e-commerce-website-place-order

UI Screenshot-

Screenshot from 2021-04-06 17-10-39
Screenshot from 2021-04-06 17-10-49
Screenshot from 2021-04-06 17-13-48
Screenshot from 2021-04-06 17-13-55

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    @order = Order.new
+  end
+  
+  def create
+    @order = Order.new(order_params)
+    flag = true
+    @cart.cart_items.each do |item|
+      @order.cart_items << item
+      item.cart_id = nil
+      product = Product.find_by(id: item.product_id)
```
Comment (kalpana19sharma) [inline app/controllers/orders_controller.rb:18]: Use association

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      if product.quantity < item.quantity 
+        flag = false
+        flash[:notice] = "OOP'S, #{product.name} IS OUT OF STOCK!!"
+        break
+      end
+    end
+    if flag
+      @cart.cart_items.each do |item|
```
Comment (kalpana19sharma) [inline app/controllers/orders_controller.rb:26]: Do we really need to run this loop 2 times?

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    end  
+    redirect_to user_path(current_user)
+  end
+
+  private
+  def order_params
+    params.require(:order).permit(:status, :user_id, :address_id, :total)
+  end
+  
+  private
```
Comment (kalpana19sharma) [inline app/controllers/orders_controller.rb:42]: No need to specify `private` keyword multiple times

File: app/views/orders/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<h2>All Orders</h2>
+<%= "Name: #{current_user.full_name}" %>
+<%= "Email: #{current_user.email}" %>
+<% @orders.each do |order| %>
+  <p>Placed Items: </p>
+  <ul>
+    <% order.cart_items.each do |item| %>
```
Comment (kalpana19sharma) [inline app/views/orders/index.html.erb:7]: Can we use eager loading for removing n+1 database queries?

File: app/views/orders/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      <li><%= item.quantity %> x <%= item.product.name %></li>
+    <% end %>
+    <% @address = Address.where(id: order.address_id) %>
```
Comment (kalpana19sharma) [inline app/views/orders/index.html.erb:10]: Use association

File: app/views/orders/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    <%= f.label :status %>
+    <%= f.text_field :status %>
+  </div>  
+  <div class = "mb-3">
+    <%= f.label :user_id %>
+    <%= f.text_field :user_id, value: current_user.id %>
+  </div>
+  <div class = "mb-3">  
+    <%= f.label :address %> 
+    <% @address = current_user.addresses.all %>
```
Comment (kalpana19sharma) [inline app/views/orders/new.html.erb:20]: This logic should be controller and variable name should be plural

File: config/routes.rb [hunk 7]
```diff
@@ -5,6 +5,7 @@
   resources :products, only: %i[index new create show]
   resources :addresses, only: %i[index new create edit update destroy]
   resources :carts
+  resources :orders
```
Comment (kalpana19sharma) [inline config/routes.rb:8]: Are we using all of the default routes?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/orders_controller.rb (hunk 2 above)
Comment (Gupta-Shivanee) [reply app/controllers/orders_controller.rb:26]: Yes ma'am, in 1st loop we are checking whether required quantity of all the products are available or not, if yes then in 2nd loop we will deduct the actual product quantity in database.

**Review by kalpana19sharma — APPROVED** [review]

**Comment by kalpana19sharma** [comment]
@Gupta-Shivanee Please resolve the conflicts and `Squash and merge` the PR

**Pull request merged**